operator/builtin/transformer/recombine/recombine.go

```diff
@@ -229,46 +255,61 @@ func (r *RecombineOperator) matchIndicatesLast() bool {
 }
 
 // addToBatch adds the current entry to the current batch of entries that will be combined
-func (r *RecombineOperator) addToBatch(_ context.Context, e *entry.Entry) {
-	if len(r.batch) >= r.maxBatchSize {
+func (r *RecombineOperator) addToBatch(ctx context.Context, e *entry.Entry, source string) {
+	if len(r.batchMap[source]) >= r.maxBatchSize {
 		r.Error("Batch size exceeds max batch size. Flushing logs that have not been recombined")
-		r.flushUncombined(context.Background())
+		err := r.flushSource(ctx, source)
+		if err != nil {
+			r.Errorf("there was error flushing combined logs %s", err)
+		}
 	}
 
-	r.batch = append(r.batch, e)
+	if _, ok := r.batchMap[source]; !ok {
+		r.batchMap[source] = []*entry.Entry{e}
+	} else {
+		r.batchMap[source] = append(r.batchMap[source], e)
+	}
+
+	if len(r.batchMap) > r.maxSources {
+		r.Error("Batched source exceeds max source size. Flushing all batched logs. Consider increasing max_sources parameter")
+		r.flushUncombined(ctx)
+	}
```

Comment:
There are a couple opportunities to improve performance here, just by reorganizing the code.
1. We only need to check `if len(r.batchMap[source]) > r.maxBatchSize` after adding to an existing batch. Also, by checking this immediately _after_ adding to a batch, we will flush as soon as we hit the limit, rather than wait until next cycle.
2. We only need to check `if len(r.batchMap) > r.maxSources` after initializing a new batch.

Something like:
```go
if _, ok := r.batchMap[source]; !ok {
	r.batchMap[source] = []*entry.Entry{e}
	if len(r.batchMap) > r.maxSources {
		r.Error("Batched source exceeds max source size. Flushing all batched logs. Consider increasing max_sources parameter")
		r.flushUncombined(ctx)
	}
	return
}

r.batchMap[source] = append(r.batchMap[source], e)
if len(r.batchMap[source]) > r.maxBatchSize {	
	if err := r.flushSource(ctx, source); err != nil {
		r.Errorf("there was error flushing combined logs %s", err)
	}
}
```

Reply:
awesome find!